### ISSUE TITLE
PPO2 selects random actions on the test run

### ISSUE THREAD

**Issue by nahrak** [openai/baselines#531]
Hi!

I've just noticed that code in run.py selects random actions from the same state. I ran a simple training example like this:
_run.py --env SeaquestNoFrameskip-v0 --network cnn --num_env 1 --num_timesteps 10000 --play_

The code under play takes random actions:
_actions = model.step(obs)[0]_

It can be reproduced, if you call _model.step(obs)[0]_ several times with the same _obs_. I believe the code needs to take a deterministic actions, or maybe we need another method to make deterministic actions, because some tasks require it. I've checked and method _CategoricalPd.mode_ is not being called anywhere during a run.

**Comment by wookayin** [comment]
I think you got something wrong. It should be a stochastic policy, so the behavior is correct. Policy model assumes a stochastic distribution `pi(a | s)`, [see here](https://github.com/openai/baselines/blob/master/baselines/common/policies.py#L50).

If you need a deterministic policy, consider fixing a random seed.

**Comment by nahrak** [comment]
It is a stochastic policy during training, but it needs a deterministic evaluation during test. Nobody would trust a self driving car taking stochastic actions. I believe PPO1 has this evaluation available in code.

**Comment by rkojcev** [comment]
This is good point @nahrak can someone tell us how to have deterministic evaluation during test?

**Comment by pzhokhov** [comment]
Hi @nahrak! IMHO the argument as to whether policy needs to be stochastic or deterministic during test is somewhat philosophical; but it should be up to the user / operator to enable that. There is no option available for that in ppo2, but we can add that. 
In the mean time, you can use a somewhat hacky way of replacing `pi.step(observation)` with 
`pi._evaluate(pi.pd.mode(), observation)` for non-recurrent policies and 
`pi.step(observation, S=state, M=mask)` with `pi._evaluate(pi.pd.mode(), observation, S=state, M=mask)` for recurrent policies in your evaluation code. Hope that helps!

**Comment by nahrak** [comment]
Hi!

In some cases doing stochastic actions is good, for example, Atari arcade games look much better (I wonder if DOTA bot use stochastic actions too?). But in other situations this might be really bad. What I can think of are self driving cars, financial decision applications and games like chess. So having an option for end users to choose from should be good.

I've already changed step() to also evaluate pd.logits and pd.mode(), although only one of these is enough for deterministic actions. This was not hard at all.

**Comment by rkojcev** [comment]
@pzhokhov will this feature be added soon in the code for PPO2? Honestly PPO1 is good in this sense because you can decide if you want to run the policy with deterministic or stochastic test.

I am are using the mlp policy with PPO2 and tried to change the step function as @nahrak suggested but still got the same behaviour. 

Trying to run saved network like this (the usual way):
```

make_model = lambda : ppo2.Model(policy=policy, ob_space=ob_space, ac_space=ac_space, nbatch_act=nenvs, nbatch_train=nbatch_train,
                nsteps=nsteps, ent_coef=ent_coef, vf_coef=vf_coef,
                max_grad_norm=max_grad_norm)

model = make_model()
if load_path is not None:
    print("Loading model from: ", load_path)
    model.load(load_path)

obs = env.reset()

while True:
    actions = model.step(obs)[0]
    obs, reward, done, _  = env.step(actions)
```

Any suggestions how to change it from here to run it with deterministic actions?

**Comment by YueErro** [comment]
Hi!

I am trying to get deterministic actions to run the mlp policy with PPO2 and after making the changes @pzhokhov and @nahrak suggested i also got the same behaviour @rkojcev.

The change i made is the following:
- [policies.py](https://github.com/openai/baselines/blob/master/baselines/common/policies.py#L93)
```diff
-     a, v, state, neglogp = self._evaluate([self.action, self.vf, self.state, self.neglogp], observation, *extra_feed)
+     a, v, state, neglogp = self._evaluate([self.mode, self.vf, self.state, self.neglogp], observation, **extra_feed)
```
The distribution i am using is DiagGaussianPd and not CategoricalPd, how can i run the policy with deterministic test?

**Comment by pzhokhov** [comment]
@rkojcev @YueErro could you provide more details on what are you seeing? Does the model select different actions given the same observation (given the changes of self.action -> self.mode in model.step)? 
Or the runs are not reproducible? The latter case may be because of the environment may have some randomness. 

Also, @YueErro, I would rather recommend creating a new function that calls model._evaluate(model.mode, observation) - with your changes behavior is deterministic for both train and test times, so likely training will not work very well.

**Comment by YueErro** [comment]
@pzhokhov thanks for your reply and sorry to get back to you after some time.

I have followed your suggestions and I was able to launch a deterministic evaluation of PPO2.
To contribute back to the community these are the changes I have made:

1. Created a new step function that is only called during evaluation:

```python
def step_deterministic(self, observation, **extra_feed):
        a, v, state, neglogp = self._evaluate([self.mode, self.vf, self.state, self.neglogp], observation, **extra_feed)
        if state.size == 0:
            state = None
        return a, v, state, neglogp
```

Notice the change of `_evaluate ` where instead of calling `self.action` i call `self.mode`

Then in the run script where we evaluate the policy instead of `step`, we call `step_deterministic`

2. Added a new boolean variable to [dummy_vec_env.py](https://github.com/openai/baselines/blob/master/baselines/common/vec_env/dummy_vec_env.py) in order not to reset the environment when `env.step(actions)` is done in runtime.

```diff
+   self.runtime = None

-   def step_async(self, actions)
+   def step_async(self, actions, runtime=False):
+       self.runtime = runtime
        listify = True
        try:
            if len(actions) == self.num_envs:
                listify = False
        except TypeError:
            pass

        if not listify:
            self.actions = actions
        else:
            assert self.num_envs == 1, "actions {} is either not a list or has a wrong size - cannot match to {} environments".format(actions, self.num_envs)
            self.actions = [actions]

    def step_wait(self):
        for e in range(self.num_envs):
            action = self.actions[e]
            if isinstance(self.envs[e].action_space, spaces.Discrete):
                action = int(action)

            obs, self.buf_rews[e], self.buf_dones[e], self.buf_infos[e] = self.envs[e].step(action)
-           if self.buf_dones[e]
+           if self.buf_dones[e] and not self.runtime:
                obs = self.envs[e].reset()
            self._save_obs(e, obs)
        return (np.copy(self._obs_from_buf()), np.copy(self.buf_rews), np.copy(self.buf_dones),
                self.buf_infos.copy())
```

Thinking of creating a new function `step_runtime(self, actions)` rather than changing the current one so to avoid conflicts when i rebase baselines.

**Comment by mitar** [comment]
My use case is slightly different. I would want that for a given initial observation, I get `k` best sequences of actions the model has learned. So the best sequence is the one where at every step you pick the action with highest probability of highest total reward. I think this is what people in this thread see as "deterministic test run". But I would also like to get additional `k-1` slightly worse action sequences, so in some way sequences with second-best overall probability to be successful (as seen by the model) and so on. In some way, I would like to sample, but sample only the sequences of steps with highest probability for highest reward as determined by the model.

**Comment by pzhokhov** [comment]
@mitar I assume you are talking about discrete action spaces case. In that case, you can reach into the  CategoricalPd class (categorical probability distribution), and use logits from it (so that at every step you can rank the actions and select either best, or second best etc). 

@YueErro I like the changes to policy class, but I am not sure I understand the purpose of the second set of changes (disable resetting of sub-environments in dummy_vec_env). Could you please clarify?

**Comment by YueErro** [comment]
@pzhokhov i made that second set of changes in order not to restart the environment (i am using my own one). My step function is set in a way that the environment restarts every time the robot achieves certain reward or the maximum episode steps. 

That said, when i use it for testing (in runtime), if it has learnt properly and achieves that certain (good) reward, it restarts first, and then it does the next action every time from then on.

What i want is the robot to get to a certain pose without going back to the start pose (restart) under any circumstances with the deterministic test.
